### PR TITLE
chore: upstream left and right tactics

### PR DESCRIPTION
**Pull request by semorrison**

### CODE REVIEW

File: Std/Tactic/LeftRight.lean [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+checking that it is an inductive type,
+and that there are the expected number of constructors.
+-/
+def nthConstructor (name : Name) (idx : Nat) (max : Option Nat) (goal : MVarId) :
+    MetaM (List MVarId) := do
+  goal.withContext do
+    goal.checkNotAssigned name
+    matchConstInduct (← goal.getType').getAppFn
+      (fun _ ↦ throwTacticEx `constructor goal "target is not an inductive datatype")
+      fun ival us ↦ do
```
Comment (digama0) [inline Std/Tactic/LeftRight.lean:23]:
```suggestion
      (fun _ => throwTacticEx `constructor goal "target is not an inductive datatype")
      fun ival us => do
```

**Comment by digama0** [comment]
needs a test

**Comment by semorrison** [comment]
Added a test.

**Comment by digama0** [comment]
also examples for inductives with one constructor, three constructors, definitions that unfold to inductives, and non-inductives

**Comment by semorrison** [comment]
Completed tests.

**Pull request merged**